Drop-in cfg files support in runtime-rs

pmores opened the pull request:
This PR adds handling of drop-in config file fragments to runtime-rs to match the same feature already implemented in golang runtime.  Behaviour of this Rust port should be the same as the Golang version so no additional documentation should be necessary.

The change is fairly non-intrusive, mostly adding new code, but the final commit does involve some refactoring of the existing code.  Admittedly there were multiple ways to achieve the goal of creating a single point that drop-in handling needs to be added to.  Being not entirely clear about the intended usage of `load_raw_from_file()` (a public function, however it doesn't seem to be called from anywhere currently) and `load()` (public as well but only called internally AFAICT) I just went for what seemed to be the simplest and least intrusive option.

Tim-0731-Hzt commented:
> a public function, however it doesn't seem to be called from anywhere currently) and `load()` (public as well but only called internally

Yes, I think so. Both `load_raw_from_file()` and `load()` could be made privately.

> This PR adds handling of drop-in config file fragments to runtime-rs to match the same feature already implemented in golang runtime. Behaviour of this Rust port should be the same as the Golang version so no additional documentation should be necessary.
> 
> The change is fairly non-intrusive, mostly adding new code, but the final commit does involve some refactoring of the existing code. Admittedly there were multiple ways to achieve the goal of creating a single point that drop-in handling needs to be added to. Being not entirely clear about the intended usage of `load_raw_from_file()` (a public function, however it doesn't seem to be called from anywhere currently) and `load()` (public as well but only called internally AFAICT) I just went for what seemed to be the simplest and least intrusive option.

Yes, I think so. Both `load_raw_from_file()` and `load` could be made privately.

pmores commented on src/libs/kata-types/src/config/mod.rs:
```diff
@@ -94,21 +94,15 @@ impl TomlConfig {
-        let config = Self::load(&content)?;
+        let mut result = Self::load_raw_from_file(config_file);
+        if let Ok((ref mut config, _)) = result {
+            Hypervisor::adjust_config(config)?;
```
I believe it should be, fundamentally.  However, `load()` is used to initialise `KATA_DEFAULT_CONFIG` and `KATA_ACTIVE_CONFIG` and I wasn't sure what the ramifications would be if I messed with `load()`'s implementation.  

Also, if `load()` were to handle drop-ins as well its interface would need to change - it would have to take a _path_ to the base config file, not just its _contents_ as it's impossible to figure out where to look for drop-ins given just the base config file contents.

That said I'm very open to suggestions how best to integrate drop-in handling into existing config loading code.  I'm not at all sure the path I've taken is the best one.

Tim-0731-Hzt replied:
The implementation looks good so far

Tim-0731-Hzt commented on src/libs/kata-types/src/config/mod.rs:
```diff
@@ -164,6 +160,267 @@ impl TomlConfig {
     }
 }
 
+mod drop_in {
```
I think it's better to put the `drop_in` mod and the unit test into a new file `drop_in.rs` under the `config` folder, and declare it  in this file.

pmores replied:
I agree and will do!

pmores commented:
@Tim-0731-Hzt I pushed a new version that moves the drop-in handling implementation out to `drop_in.rs` as suggested.  This is just a pure code reorganisation, none of the actual algorithms are changed.

I also took this opportunity to split the private implementation into two private submodules to make it explicit that it has two distinct layers.  The lower one deals exclusively with TOML parse trees merging and is used by the higher one that handles the filesystem-related aspect of drop-in cfg file fragments.  The public reexport of `load()` is the only public interface of the `drop_in` module.

quanweiZhou commented:
/test-dragonball

pmores commented:
@quanweiZhou I'm looking into the dragonball test failure.

pmores commented:
@quanweiZhou Wow, there was an actual omission in `config.d` handling - good catch, CI, and shame on me! :-) I fixed that and added a couple of related tests.

pmores commented:
/test-dragonball

liubin commented:
Since `runtime-rs` has been merged, please change your target branch to `main`.

pmores commented:
@liubin Thanks for reminding me, done!

liubin commented:
/test

liubin commented:
/test-arm

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -0,0 +1,215 @@
+                    string: "foo".into(),
+                    sub: SubConfig {
+                        ..Default::default()
+                    },
```
Can just use `sub: Default::default()`

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -213,3 +213,205 @@ mod toml_tree_ops {
+    fn get_dropin_dir_path(base_cfg_file_path: &Path) -> Result<PathBuf> {
+        let mut dropin_dir = base_cfg_file_path.to_path_buf();
+        if !dropin_dir.pop() {
+            return Err(io::Error::new(
```
Using `io::Error` here is maybe a bit odd, since the error isn't really related to IO.  Consider using `anyhow::Error` instead?

dgibson replied:
Ok, fair enough.

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -213,3 +213,205 @@ mod toml_tree_ops {
+    }
+
+    fn update_from_dropin(base_config: &mut toml::Value, dropin_file: &fs::DirEntry) -> Result<()> {
+        if !dropin_file.file_type()?.is_file() && !dropin_file.file_type()?.is_symlink() {
```
Double check if you need `is_symlink`; in many languages a symlink to a regular file will pass both "is file" and "is symlink" checks.

pmores replied:
Yes, you're right and this is an actual bug - good catch, thanks for that! A symlink to a directory would be handled inappropriately by this code.

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -213,3 +213,205 @@ mod toml_tree_ops {
+
+        let mut dropin_files = dropin_files_iter.collect::<Result<Vec<_>>>()?;
+        dropin_files.sort_by_key(|direntry| direntry.file_name());
+        for dropin_file in dropin_files.iter() {
```
Can be just `for dropin_file in &dropin_files`

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -213,3 +213,205 @@ mod toml_tree_ops {
+                internetworking_model="tcfilter"
+            "#;
+
+            let mut config_path = PathBuf::from(tmpdir.path());
+            config_path.push("runtime.toml");
```
Slightly clearer to use `tmpdir.path().join("runtime.toml")`

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -213,3 +213,205 @@ mod toml_tree_ops {
+            let mut config_path = PathBuf::from(tmpdir.path());
+            config_path.push("runtime.toml");
+            let result = create_file(&config_path, runtime_config_file_data.as_bytes());
+            assert!(result.is_ok());
```
Could just `unwrap()` in the previous line - unwrap is a special form of assertion, so is reasonable in tests.

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -213,3 +213,205 @@ mod toml_tree_ops {
+            );
+            assert_eq!(config.hypervisor["qemu"].device_info.default_bridges, 3);
+            assert!(config.runtime.debug);
+            assert_eq!(config.runtime.internetworking_model, "tcfilter".to_string());
```
Could use a common helper function for the last piece of the test in both `test_no_dropins_dir` and `test_no_dropins`.

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -213,3 +213,205 @@ mod toml_tree_ops {
+        fn test_no_dropins() {
+            let tmpdir = tempfile::tempdir().unwrap();
+
+            let runtime_config_file_data = r#"
+                [hypervisor.qemu]
+                path = "/usr/bin/qemu-kvm"
+                default_bridges = 3
+                [runtime]
+                enable_debug = true
+                internetworking_model="tcfilter"
+            "#;
```
Could use a common const definition for this in the test module.

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -213,3 +213,205 @@ mod toml_tree_ops {
+            );
+            assert_eq!(config.hypervisor["qemu"].cpu_info.default_vcpus, 2);
+            assert_eq!(config.hypervisor["qemu"].device_info.default_bridges, 4);
+            assert_eq!(
+                config.hypervisor["qemu"].shared_fs.shared_fs,
+                Some("virtio-9p".to_string())
+            );
```
This does an allocation of a new copy of the "virtio-9p" string, which is a little bit ugly, though it doesn't really matter for a test case.  If you wanted to, the way to do this comparison without the allocation is:
```
assert_eq!(config.hypervisor["qemu"].shared_fs.shared_fs.as_ref(), Some("virtio-9p"));
```

dgibson reviewed (approved):
I have commented with a number of minor nits that could be polished, but overall this looks good to me.

pmores commented:
Pushed a new version to integrate @dgibson's proposals.  Thank you very much for taking the time to do such a detailed review.

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -212,3 +212,171 @@ mod toml_tree_ops {
+            let dropin_dir = tmpdir.path().join("config.d");
+            fs::create_dir(&dropin_dir).unwrap();
+
+            let dropin_path = dropin_dir.as_path().join("10-base");
```
I don't believe you need the `.as_path()` calls - Deref coercion should mean you can can call `join` directly on the `PathBuf`.

dgibson commented on src/libs/kata-types/src/config/drop_in.rs:
```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2022 Red Hat
```
Not really an upstream consideration, but fwiw, I believe Red Hat Legal recommends using simply `Copyright Red Hat` rather than putting years or `(c)` in such notices (I was surprised too).

dgibson reviewed (approved):
Thanks for the polish changes.  I noted a couple of remaining trivial nits, but LGTM.

jodh-intel reviewed (approved):
Thanks @pmores.

lgtm, although we will need some doc to support this change. Could you tal at that in a follow-up PR?

pmores commented:
Pushed the remaining fixes as proposed by @dgibson.

@jodh-intel Thanks! Obviously I'm happy to adjust docs but I'm afraid I'll need a hint as to what's necessary. I thought originally that existing documentation would do since this change is meant to work precisely the same as the golang version but I'm probably overlooking something.

liubin commented:
/test
